### PR TITLE
Fix points data selection for 4D

### LINKED ISSUE #6680: Select Mode of Points Layer broken in 4 dimensions
### 🐛 Bug Report

When displaying points 4D with the 3D view, selecting the "Select" mode of the Points layer then dragging the mouse across the canvas (not moving point, just dragging the cursor as if to create the usual selection bounding box) does nothing and raises an Assertion error. If a point is clicked instead, it can be moved normally without any error.
Here's the error I get
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
File ~/anaconda3/envs/napari-dev/lib/python3.9/site-packages/superqt/utils/_throttler.py:249, in ThrottledCallable._set_future_result(self=<superqt.utils._throttler.ThrottledCallable object>)
    248 def _set_future_result(self):
--> 249     result = self._func(*self._args[: self._max_args], **self._kwargs)
        self._func = <bound method VispyCanvas._on_mouse_move of <napari._vispy.canvas.VispyCanvas object at 0x7ff6c544fdf0>>
        self = <superqt.utils._throttler.ThrottledCallable object at 0x7ff6ac3160d0>
        self._args[: self._max_args] = (<MouseEvent blocked=False button=1 buttons=[1, 1] delta=[0. 0.] handled=False is_dragging=True last_event=MouseEvent modifiers=() native=<PyQt5.QtGui.QMouseEvent object at 0x7ff6c545c310> pos=[527 398] press_event=MouseEvent source=<NapariSceneCanvas (PyQt5) at 0x7ff6c5459070> sources=[<NapariSceneCanvas (PyQt5) at 0x7ff6c5459070>] time=1708521467.3701262 type=mouse_move>,)
        self._kwargs = {}
        self._args = (<MouseEvent blocked=False button=1 buttons=[1, 1] delta=[0. 0.] handled=False is_dragging=True last_event=MouseEvent modifiers=() native=<PyQt5.QtGui.QMouseEvent object at 0x7ff6c545c310> pos=[527 398] press_event=MouseEvent source=<NapariSceneCanvas (PyQt5) at 0x7ff6c5459070> sources=[<NapariSceneCanvas (PyQt5) at 0x7ff6c5459070>] time=1708521467.3701262 type=mouse_move>,)
        self._max_args = 1
    250     self._future.set_result(result)

File /napari/napari/_vispy/canvas.py:448, in VispyCanvas._on_mouse_move(self=<napari._vispy.canvas.VispyCanvas object>, event=<MouseEvent blocked=False button=1 buttons=[1, 1...459070>] time=1708521467.3701262 type=mouse_move>)
    436 def _on_mouse_move(self, event: MouseEvent) -> None:
    437     """Called whenever mouse moves over canvas.
    438 
    439     Parameters
   (...)
    446     None
    447     """
--> 448     self._process_mouse_event(mouse_move_callbacks, event)
        event = <MouseEvent blocked=False button=1 buttons=[1, 1] delta=[0. 0.] handled=False is_dragging=True last_event=MouseEvent modifiers=() native=<PyQt5.QtGui.QMouseEvent object at 0x7ff6c545c310> pos=[527 398] press_event=MouseEvent source=<NapariSceneCanvas (PyQt5) at 0x7ff6c5459070> sources=[<NapariSceneCanvas (PyQt5) at 0x7ff6c5459070>] time=1708521467.3701262 type=mouse_move>
        self = <napari._vispy.canvas.VispyCanvas object at 0x7ff6c544fdf0>

File /napari/napari/_vispy/canvas.py:409, in VispyCanvas._process_mouse_event(self=<napari._vispy.canvas.VispyCanvas object>, mouse_callbacks=<function mouse_move_callbacks>, event=<ReadOnlyWrapper at 0x7ff69675b680 for MouseEvent>)
    407 layer = self.viewer.layers.selection.active
    408 if layer is not None:
--> 409     mouse_callbacks(layer, event)
        event = <ReadOnlyWrapper at 0x7ff69675b680 for MouseEvent at 0x7ff6954096d0>
        layer = <Points layer 'points' at 0x7ff6eb790c40>
        mouse_callbacks = <function mouse_move_callbacks at 0x7ff6debac940>

File /napari/napari/utils/interactions.py:178, in mouse_move_callbacks(obj=<Points layer 'points'>, event=<ReadOnlyWrapper at 0x7ff69675b680 for MouseEvent>)
    175 obj._persisted_mouse_event[gen].__wrapped__ = event
    176 try:
    177     # try to advance the generator
--> 178     next(gen)
        gen = <generator object select at 0x7ff6978be890>
    179 except StopIteration:
    180     # If done deleted the generator and stored event
    181     del obj._mouse_drag_gen[func]

File /napari/napari/layers/points/_points_mouse_bindings.py:92, in select(layer=<Points layer 'points'>, event=<ReadOnlyWrapper at 0x7ff696756100 for ReadOnlyWrapper>)
     89     layer._drag_box = np.array([layer._drag_start, coord])
     91     # update the drag up and normal vectors on the layer
---> 92     _update_drag_vectors_from_event(layer=layer, event=event)
        layer = <Points layer 'points' at 0x7ff6eb790c40>
        event = <ReadOnlyWrapper at 0x7ff696756100 for ReadOnlyWrapper at 0x7ff69675b680>
     94     layer._set_highlight()
     95 yield

File /napari/napari/layers/points/_points_mouse_bindings.py:195, in _update_drag_vectors_from_event(layer=<Points layer 'points'>, event=<ReadOnlyWrapper at 0x7ff696756100 for ReadOnlyWrapper>)
    191 if n_display == 3:
    192     # if in 3D, set the drag normal and up directions
    193     # get the indices of the displayed dimensions
    194     ndim_world = len(event.position)
--> 195     layer_dims_displayed = layer._world_to_layer_dims(
        layer = <Points layer 'points' at 0x7ff6eb790c40>
        ndim_world = 4
        event = <ReadOnlyWrapper at 0x7ff696756100 for ReadOnlyWrapper at 0x7ff69675b680>
    196         world_dims=event.dims_displayed, ndim_world=ndim_world
    197     )
    199     # get the view direction in displayed data coordinates
    200     layer._drag_normal = layer._world_to_displayed_data_ray(
    201         event.view_direction, layer_dims_displayed
    202     )

File /napari/napari/layers/base/base.py:1640, in Layer._world_to_layer_dims(self=<Points layer 'points'>, world_dims=[1, 2, 3], ndim_world=4)
   1580 def _world_to_layer_dims(
   1581     self, *, world_dims: npt.NDArray, ndim_world: int
   1582 ) -> np.ndarray:
   1583     """Map world dimensions to layer dimensions while maintaining order.
   1584 
   1585     This is used to map dimensions from the full world space defined by ``Dims``
   (...)
   1638         The corresponding layer dimensions with the same ordering as the given world dimensions.
   1639     """
-> 1640     return self._world_to_layer_dims_impl(
        self = <Points layer 'points' at 0x7ff6eb790c40>
        world_dims = [1, 2, 3]
        ndim_world = 4
   1641         world_dims, ndim_world, self.ndim
   1642     )

File /napari/napari/layers/base/base.py:1652, in Layer._world_to_layer_dims_impl(world_dims=<class 'numpy.ndarray'> (3,) int64, ndim_world=4, ndim=4)
   1648 """
   1649 Static for ease of testing
   1650 """
   1651 world_dims = np.asarray(world_dims)
-> 1652 assert world_dims.min() == 0
        world_dims = <class 'numpy.ndarray'> (3,) int64
   1653 assert world_dims.max() == len(world_dims) - 1
   1654 assert world_dims.ndim == 1

AssertionError:
```

When the cursor is not dragged but clicked, an Axis error raises (`axisa: axis -1 is out of bounds for array of dimension 0`). I won't show the error stack as it is quite long.

### 💡 Steps to Reproduce

1. Display points in 4D. You can use the following snippet:
```python
import napari
import numpy as np

np.random.seed(0)

data = np.random.random((2, 100,100,100))
points = np.random.random((3, 4)) * 100
points[:,0] = 0

viewer = napari.Viewer(ndisplay=3)
viewer.add_image(data)
viewer.add_points(points, size=10)
napari.run()
```
2. Go in 3D view mode
3. Make sure the "Select" mode of the Points Layer is selected. Click and drag your cursor across the canvas. 

### 💡 Expected Behavior

Normally, the action described should draw a thin blue rectangular bounding box that allows selecting the points.

### 🌎 Environment

napari: 0.1.dev3244+g0747da5
Platform: Linux-5.15.0-91-generic-x86_64-with-glibc2.31
System: Ubuntu 20.04.6 LTS
Python: 3.9.0 (default, Nov 15 2020, 14:28:56)  [GCC 7.3.0]
Qt: 5.15.2
PyQt5: 5.15.9
NumPy: 1.24.2
SciPy: 1.9.1
Dask: 2023.4.0
VisPy: 0.14.1
magicgui: 0.7.2
superqt: 0.6.1
in-n-out: 0.1.7
app-model: 0.2.4
npe2: 0.7.4

OpenGL:
  - GL version:  4.6 (Compatibility Profile) Mesa 21.2.6
  - MAX_TEXTURE_SIZE: 16384
  - GL_MAX_3D_TEXTURE_SIZE: 2048

Settings path:
  - /home/jvanaret/.config/napari/napari-dev_ebc9c46b35c7ad556420cc56947aae90c9e81949/settings.yaml
Plugins:
  - napari: 0.1.dev3244+g0747da5 (81 contributions)
  - napari-console: 0.0.9 (0 contributions)
  - napari-svg: 0.1.10 (2 contributions)
  - napari-threedee: 0.0.12 (28 contributions)

### 💡 Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dalthviz**
# References and relevant issues

Closes #6680 

# Description

Fix handling multiple points selection for 4D. So, with the following example snippet:

```python
import napari
import numpy as np

np.random.seed(0)

data = np.random.random((2, 100,100,100))
points = np.random.random((3, 4)) * 100
points[:,0] = 0

viewer = napari.Viewer(ndisplay=3)
viewer.add_image(data)
viewer.add_points(points, size=10)
napari.run()
```

Being able to do something like:

points_4d

Also, extends some existing tests: one for points selection to cover a check for 4D and another one that checks `_world_to_layer_dims_impl` adding the params that are passed when doing a selection like the illustrated above

### CODE REVIEW

File: napari/layers/base/base.py [hunk 1]
```diff
@@ -1659,14 +1659,14 @@ def _world_to_layer_dims_impl(
...
-        assert world_dims.ndim == 1
         offset = ndim_world - ndim
-        order = world_dims - offset
-        order = order[order >= 0]
-        return order - order.min()
+        order = np.array(world_dims)
+        if offset == 0:
+            return order
+        if offset < 0:
+            return np.concatenate((np.arange(-offset), order - offset))
```
Comment (brisvag) [inline napari/layers/base/base.py:1832]: A comment here would be great, explaining what offset is, I find this  abit hard to follow :P

**Review by brisvag — APPROVED** [review]

File: napari/layers/base/base.py (hunk 1 above)
Comment (dalthviz) [reply napari/layers/base/base.py:1832]: Thanks for giving this a check! While working on the comment and rechecking things locally I noticed that this fails when trying to do the point selection after changing the viewer to 2D :( Will mark this as a draft for the moment and keep checking.

Edit: Just in case, the error I'm seeing is something like

```python-traceback
Traceback (most recent call last):
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\app\backends\_qt.py", line 928, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\util\event.py", line 453, in __call__
    self._invoke_callback(cb, event)
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
    _handle_exception(self.ignore_callback_errors,
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\util\event.py", line 469, in _invoke_callback
    cb(event)
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\scene\canvas.py", line 219, in on_draw
    self._draw_scene()
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\scene\canvas.py", line 278, in _draw_scene
    self.draw_visual(self.scene)
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\scene\canvas.py", line 316, in draw_visual
    node.draw()
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\scene\visuals.py", line 106, in draw
    self._visual_superclass.draw(self)
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\visuals\visual.py", line 668, in draw
    v.draw()
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\scene\visuals.py", line 106, in draw
    self._visual_superclass.draw(self)
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\visuals\visual.py", line 668, in draw
    v.draw()
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\visuals\visual.py", line 505, in draw
    if self._prepare_draw(view=self) is False:
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\visuals\line\line.py", line 331, in _prepare_draw
    self._pos_vbo.set_data(pos)
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\gloo\buffer.py", line 189, in set_data
    data = self._prepare_data(data, **kwargs)
  File "C:\Users\dalth\anaconda3\envs\napari-dev\lib\site-packages\vispy\gloo\buffer.py", line 446, in _prepare_data
    raise ValueError('Last dimension should be %s not %s'
ValueError: Last dimension should be 3 not 2
```

**Comment by brisvag** [comment]
Hey @dalthviz, just got a notification from this! Are stuck waiting for more feedback, or is it still on hold on your side?

**Comment by dalthviz** [comment]
Hi there @brisvag ! And just updating this will recheking it but seems like I'm still facing the error (https://github.com/napari/napari/pull/6819#discussion_r1562964127) so this is still on hold but thanks for reaching out to check!

**Comment by brisvag** [comment]
Ok! Let me know if you'd like some help with it :)

**Comment by dalthviz** [comment]
So I gave another try to this but no clue where I could do changes to prevent the error 😅 From my testing seems like the first attempt to select points (regardless of the viewer being in 2d or in 3d mode) works. However, once you switch viewer dimension mode (2d to 3d or 3d to 2d) the error mentioned above appears:

* From 3d to 2d:

points_selection_4d

* From 2d to 3d:

points_selection_4d_2d

My guess is that when the viewer changes dimensionality a change over vispy related code is missing (like an update of some sort? 🤔) Anyhow, if you have some time/want to try pushing this forward @brisvag feel free to do any changes over this PR branch :)

**Comment by brisvag** [comment]
@dalthviz Foudn the problem! Not sure why, but I can't seem to be able to push, so I opened a pr to your pr instead! https://github.com/dalthviz/napari/pull/9
